caluma/permissions.py

```diff
@@ -1,36 +1,87 @@
+
+        _decorated._permission = mutation
+        return _decorated
+
+    return decorate
+
+
+def object_permission_for(mutation):
+    """Decorate function to overwriting object permission of specific mutation."""
+
+    def decorate(fn):
+        @wraps(fn)
+        def _decorated(self, mutation, info, instance):
+            return fn(self, mutation, info, instance)
+
+        _decorated._object_permission = mutation
+        return _decorated
+
+    return decorate
 
 
 class BasePermission(object):
-    """Basic permission class to be extended by any permission implementation."""
+    """Basic permission class to be extended by any permission implementation.
 
-    def has_permission(self, mutation, info):
-        """
-        Check permission of specific mutation or otherwise return True.
+    In combination with the decorators `@permission_for` and `@object_permission_for` a custom
+    permission class can define permission on basis of mutations and its subclasses.
 
-        Checks for mutation specific implementations by its snake name.
+    Per default it returns True but default can be overwritten to define a permission for
+    `Mutation`
 
-        e.g. for mutation SaveDocument it will call `has_permission_for_save_document` if available.
-        """
-        method_name = f"has_permission_for_{to_snake_case(mutation.__name__)}"
-        has_permission_for_mutation = getattr(self, method_name, None)
-        if has_permission_for_mutation is not None:
-            return has_permission_for_mutation(mutation, info)
+    A custom permission class could look like this:
+    ```
+    >>> from caluma.form.schema import SaveForm
+    ... from caluma.mutation import Mutation
+    ...
+    ... class CustomPermission(BasePermission):
+    ...     @permission_for(Mutation)
+    ...     def has_permission_default(self, mutation, info):
+    ...         # change default permission to False when no more specifc
+    ...         # permission is defined.
+    ...         return False
+    ...
+    ...     @permission_for(SaveForm)
+    ...     def has_permission_for_save_form(self, mutation, info):
+    ...         return True
+    ...
+    ...     @object_permission_for(SaveForm)
+    ...     def has_object_permission_for_save_form(self, mutation, info, instance):
+    ...         return instance.slug != 'protected-form'
+    """
 
-        return True
+    def __init__(self):
+        perm_fns = inspect.getmembers(self, lambda m: hasattr(m, "_permission"))
+        self._permissions = {fn._permission: fn for _, fn in perm_fns}
 
-    def has_object_permission(self, mutation, info, instance):
-        """
-        Check object permission of specific mutation or otherwise return True.
+        object_perm_fns = inspect.getmembers(
+            self, lambda m: hasattr(m, "_object_permission")
+        )
+        self._object_permissions = {
+            fn._object_permission: fn for _, fn in object_perm_fns
+        }
```

Comment:
It'd probably be a good idea to safeguard against duplicate permission methods here. Throw an exception if two methods are tagged for the same permission class

Reply:
Added safeguarding